monai/apps/auto3d/data_analyzer.py

```diff
@@ -0,0 +1,618 @@
+                {"fold": 1, "image": "image_004.nii.gz", "label": "label_004.nii.gz"},
+            ],
+        }
+
+        dataroot = '/datasets' # the directory where you have the image files (nii.gz)
+        DataAnalyzer(datalist, dataroot)
+
+    Note:
+        The module can also be called from the command line interface (CLI) if MONAI is
+        installed with the source code.
+
+    For example:
+
+    .. code-block:: bash
+
+        python monai/apps/auto3d \
+            DataAnalyzer \
+            get_all_case_stats \
+            --datalist="my_datalist.json" \
+            --dataroot="my_dataroot_dir"
+
+    """
+
+    def __init__(
+        self,
+        datalist: Union[str, Dict],
+        dataroot: str,
+        output_path: str = "./data_stats.yaml",
+        average: bool = False,
+        do_ccp: bool = True,
+        device: Union[str, torch.device] = "cuda",
+        worker: int = 2,
+    ):
+        """
+        The initializer will load the data and register the functions for data statistics gathering.
+        """
+        self.output_path = output_path
+        files, _ = datafold_read(datalist=datalist, basedir=dataroot, fold=-1)
+        ds = data.Dataset(
+            data=files,
+            transform=transforms.Compose(
+                [
+                    transforms.LoadImaged(keys=["image", "label"]),
+                    transforms.EnsureChannelFirstd(keys=["image", "label"]),  # this creates label to be (1,H,W,D)
+                    transforms.Orientationd(keys=["image", "label"], axcodes="RAS"),
+                    transforms.EnsureTyped(keys=["image", "label"], data_type="tensor"),
+                    transforms.Lambdad(
+                        keys="label", func=lambda x: torch.argmax(x, dim=0, keepdim=True) if x.shape[0] > 1 else x
+                    ),
+                    transforms.SqueezeDimd(keys=["label"], dim=0),  # make label (H,W,D)
+                ]
+            ),
+        )
+        self.dataset = data.DataLoader(ds, batch_size=1, shuffle=False, num_workers=worker, collate_fn=no_collation)
+        # Whether to average all the modalities in the summary
+        # If SUMMARY_AVERAGE is set to false,
+        # the stats for each modality are calculated separately
+        self.SUMMARY_AVERAGE = average
+        self.DO_CONNECTED_COMP = do_ccp
+        self.device = device
+        self.data: Dict[str, Any] = {}
+        self.results: Dict[str, Any] = {}
+        # gather all summary function for combining case stats
+        self.gather_summary: Dict[str, Dict] = {}
+        self._register_functions()
+        self._register_operations()
+
+    def _register_functions(self):
+        """
+        Register all the statistic functions for calculating stats for individual cases and overall. If one installs
+        monai in the editable source code manner, then a new statistics calculation can be customized in the DataAnalyzer
```

Comment:
instead of editing the source code, would be great to support this use case if time permits:
```py
from monai.apps.auto3d import DataAnalyzer

def my_func(...):
    ...

def my_stats(...):
    ...

def my_stats_summary(...):
    ...


analyzer = DataAnalyzer(...)
analyzer.register_functions(func=my_func, stats=my_stats, stats_summary=stats_summary)
```